Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,747 @@
+That is, indexing a `[|'T|]` by `(Range, Range)` returns a rank-2 array (`[|'T|]`), while indexing by `(Int, Range)` or `(Range, Int)` returns an ordinary rank-1 array (`['T]`).
+
+Just as with indices like `(Int, Int)` and `(Int, Int, Int)`, subscripts that return slices can also be used in copy-and-update expressions, as shown in Example 6.
+Multidimensional indices can also be used with the copy-and-update operator (`w/`) to create a new array that replaces one or more specific element(s) of a multidimensional array, as shown in Example 4 below.
+In the case of `w/` operators, the `()` around index tuples cannot be dropped.
+
+When using values of type `Range` to index one or more axes in a multidimensional array the contextual `Range` expression `...`, e.g., is shorthand for the value `0..1..(n - 1)` where `n` is the length of the axis being indexed. The same [contextual expressions](https://github.com/microsoft/qsharp-language/blob/main/Specifications/Language/3_Expressions/ContextualExpressions.md#contextual-and-omitted-expressions) as for the one-dimensional case are valid for each dimension in the array. 
+
+When used in `for` loops, multidimensional arrays iterate "on the left," yielding loop variables of one rank lower than the array being looped over, as shown in Example 7, below.
+
+Finally, to support multidimensional arrays, this proposal also suggests extending the `Microsoft.Quantum.Arrays` namespace with the following functions that can be used to implement libraries for working with multidimensional arrays:
+
+- `internal function AsNDArrayUnsafe<'TElement, 'TArray>(data : ['TElement], strides : [Int], offset : Int, size : [Int]) : 'TArray`
+- `internal function NDArrayDataUnsafe<'TElement, 'TArray>(data : 'TArray) : ['TElement]`
+- `internal function NDArrayStridesUnsafe<'TArray>(data : 'TArray) : [Int]`
+- `internal function NDArraySizeUnsafe<'TArray>(data : 'TArray) : [Int]`
+- `internal function NDArrayOffsetUnsafe<'TArray>(data : 'TArray) : Int`
+
+Each of these five functions would be `body intrinsic;`, and together form the contract between the runtime and the core Q# libraries needed to support this proposed feature (see Example 7, below). By necessity, these functions are "unsafe," in that direct access to these functions would allow violating invariants of multidimensional arrays or bypass the type system to expose runtime failures, necessitating the `internal` modifier. In particular, these functions may be given access to elements that do not exist insofar as Q# user code is concerned (i.e.: are not accessible given the current strides, shapes, and offsets); these elements may have reference counts of zero, such that referring to these elements is undefined behavior. It is the responsibility of the Q# foundation, core, and standard libraries to ensure that undefined behavior is not exposed to user code.
```

Comment:
```suggestion
Each of these five functions would be `body intrinsic;`, and together form the contract between the runtime and the core Q# libraries needed to support this proposed feature (see Example 7, below). By necessity, these functions are "unsafe," in that direct access to these functions would allow violating invariants of multidimensional arrays or bypass the type system to expose runtime failures, necessitating the `internal` modifier. In particular, these functions may be given access to elements that do not exist insofar as Q# user code is concerned (i.e.: are not accessible given the current strides, shapes, and offsets); these elements may have reference counts of zero, such that referring to these elements results in undefined behavior. It is the responsibility of the Q# foundation, core, and standard libraries to ensure that undefined behavior is not exposed to user code.
```